oak_runtime/src/node/http/mod.rs

```diff
@@ -518,21 +494,131 @@ impl HttpResponseIterator {
         response_receiver.receive(&self.runtime)
     }
 
-    fn to_response(&self) -> Result<Response<Body>, http::Error> {
+    fn to_hyper_response(&self) -> Result<Response<Body>, OakStatus> {
         info!(
             "Generating response for runtime {} and reader {:?}.",
             self.runtime.node_id.0, self.response_reader
         );
         match self.read_response() {
-            Ok(http_response) => http::response::Builder::new()
-                .status(http_response.status as u16)
-                .body(Body::from(http_response.body)),
+            Ok(http_response) => to_hyper_http_response(http_response),
             Err(status) => {
-                error!("Could not read response: {}", status);
+                warn!("Could not read response: {}", status);
                 http::response::Builder::new()
                     .status(StatusCode::INTERNAL_SERVER_ERROR)
                     .body(Body::empty())
+                    .map_err(|err| {
+                        warn!("Could not build response: {}", err);
+                        OakStatus::ErrInternal
+                    })
             }
         }
     }
 }
+
+/// Wrapper around a HashMap representing the headers in protobuf encapsulated HttpRequest and
+/// HttpResponse. This struct is defined mainly to allow implementing `std::convert::From` for more
+/// convenient conversion from hyper HeaderMaps.
+struct OakHeaderMap {
+    headers: HashMap<String, HeaderValue>,
+}
+
+impl OakHeaderMap {
+    // Flatten the headers into a vector of tuples (String, http::header::HeaderValue) that can
+    // be easily iterated over when populating the headers in the hyper Response builder.
+    fn to_vec(&self) -> Vec<(http::header::HeaderName, http::header::HeaderValue)> {
+        self.headers
+            .iter()
+            .flat_map(|(name, value)| {
+                value
+                    .values
+                    .iter()
+                    .map(move |val| {
+                        let header_name = http::header::HeaderName::from_bytes(name.as_bytes())
+                            .map_err(|err| {
+                                warn!("Error when parsing header name: {}", err);
+                                err
+                            });
+                        let header_value =
+                            http::header::HeaderValue::from_bytes(val).map_err(|err| {
+                                warn!("Error when parsing header value: {}", err);
+                                err
+                            });
+                        (header_name, header_value)
+                    })
+                    .filter(|(name, value)| name.is_ok() && value.is_ok())
+                    // can safely call unwrap since the `Error`s are filtered out
+                    .map(|(name, value)| (name.unwrap(), value.unwrap()))
```

Comment:
I think this can be a single `filter_map` https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map in which the function is something like `|(name, value)| Some(name?, value?)`, assuming `?` works in lambdas, which I am not sure.

Reply:
Thanks. Done.